### ISSUE TITLE
Have a way to not boot the process per datum.

### ISSUE THREAD

**Issue by jdoliner** [pachyderm/pachyderm#1592]
Booting a process per datum can be too intensive for lesser languages such as Python and Java. There should be some way for a single process to process multiple datums.

This seems to mess with our semantics in some kind of weird ways, I'm wondering if it would be possible for services to satisfy the needs here, ie your long running process which is hit by multiple datums would be run as a service and thus spun up once and your pipeline code would just send the datum off to it, via some sort of api that the user has defined, get the data back and write it to the correct place.

**Comment by emk** [comment]
When operating on files, it would certainly be nice to have the option to get more than one file at a time, like we did in 1.3. This would allow us to do a partial reduce across several input files. I did some measurements and this reduces the size of one particular output repo by at least 30%. But we can work around the absence.

**Comment by JonathanFraser** [comment]
@emk  You can glob on a prefix. This lets you use directories to specify how many files you want available as an atomic unit.

**Comment by jdoliner** [comment]
@JonathanFraser this is a good answer to this problem, thanks for chiming in. However there is one downside to keep in mind here. If you group keys together then they will all get processed together which might be faster, but if a new file is added to the group then the whole group needs to be reprocessed which can lead to a lot of extra processing.

**Comment by JonathanFraser** [comment]
@jdoliner True....So when you have relatively small incremental commits but your initial commit is large then this will be a problem.

**Comment by gabrielgrant** [comment]
Having user code process a batch of datums (#1662) is one approach to solving this. Moving shuttling datums in and out of the user-visible `/pfs` dir is another (discussed [here](https://github.com/pachyderm/pachyderm/issues/1662#issuecomment-293980535) and [here](https://github.com/pachyderm/pachyderm/issues/1662#issuecomment-307948158) )

**Comment by OliverEvans96** [comment]
Any action here? Processing multiple datums within a single run of user code would be handy! :+1:

**Comment by JoeyZwicker** [comment]
@OliverEvans96 your best bet to solve this in the short term is going to be batch your datums at the data level so they get globbed together. If you have a workload with lots of datums and fast processing per datum where user code booting becomes a meaningful performance hit, this is going to be the simplest option for now.

**Comment by OliverEvans96** [comment]
Got it, thanks @JoeyZwicker !

**Comment by claytonlemons** [comment]
Some additional observations regarding the use of globbing/grouping to batch files into datums:

1. Depending on an application's file naming convention, a glob could be constrained to only certain grouping sizes. For example, files that have a base-10 number (e.g. `1352451.json.gz`) for a name are limited to groups of 1, 10, 100, etc., depending on the number of `?` in the glob pattern. The only way to adjust the group sizes is to change the base. Consider also the cases of file names that are random hashes or base 64 encodings. It's certainly possible to change the naming convention to make files more amenable to grouping with a glob, but an application may not always have control over the convention.
2. Related to 1, if the naming convention can't be controlled, globbing may introduce unpredictable performance patterns, as the sizes of datums may not be uniform or consistent. Continuing with the example above, datums could be sparse or dense depending on if the numeric names are randomly or sequentially assigned by an upstream system.
3. In addition to performance issues, point 2 may cause problems if a pipeline depends on lazily loaded inputs. If the datum size cannot be reliably controlled through globbing and grouping, a datum could end up with more than 10K files. Trying to balance this constraint while producing datums that are large enough to overcome boot up overhead seems challenging at best.

Proposal:
- It seems that the challenge with processing multiple datums with the same invocation of a pipeline command is that it would be unclear which files in `/pfs/out` directory would correspond with each of the datums in the input, which completely breaks the one-input-datum-one-output-datum model.
- What if the input repo directory was structured such that it was clear which files belonged to which datums. For example:
    - ```
    /pfs/datum_1/my_input/in_file_1.txt
    /pfs/datum_1/my_input/in_file_2.txt
    /pfs/datum_2/my_input/in_file_3.txt
    /pfs/datum_3/my_input/in_file_4.txt
    ```
  - `datum_*` are just placeholders for actual datum identifiers.
  - In the path, the datum identifiers come before the input repo name so that joins would still make sense.
- Now, the application code has enough of a hint to organize the output files such that it's clear which datums they belong to. For example, the following `/pfs/out` directory makes it clear that datum_1 produced 3 output files, datum_2 produced 2, and datum_3 produced none.
    ```
    /pfs/out/datum_1/out_file_1_1.txt
    /pfs/out/datum_1/out_file_1_2.txt
    /pfs/out/datum_1/out_file_2_1.txt
    /pfs/out/datum_2/out_file_3_1.txt
    /pfs/out/datum_2/out_file_3_2.txt
    ```
- To opt in to this feature, pipeline specs can specify a `batch_size` field, which will cause datums to be batched together in batches of predictable sizes. Datums could still be groups of files, but it's the onus of the application developer to understand the implications of mixing batching and grouping. The worker would ensure the datums are copied into the pfs input directory and organized so that the application would know how to separate the output files into output datums. The worker would then be able to upload the output datums as usual.

**Comment by dgeorg42** [comment]
Thanks for the feedback and proposals, @claytonlemons!  We've recently started discussing potential solutions for this particular issue, one of which is very similar to what you're suggesting.  Another solution that we're considering involves adding a "next datum" API call which could be used inside of user code.  Users could simply add a loop and a next-datum call to their code to process all of the available datums within a single execution of the user code. Existing code would continue to operate as it does today, but pipelines could be "upgraded" to use the new functionality by just modifying the code to add a loop and a next-datum call.

**Comment by claytonlemons** [comment]
@dgeorg42 That's very exciting to hear! So far, we are already seeing 70-99% efficiency in the incremental processing case for some of our pipelines, as a direct result of the provenance information that Pachyderm tracks for datums. That efficiency is only hindered by the [reprocessing](https://github.com/pachyderm/pachyderm/issues/1592#issuecomment-293752205) concern that @jdoliner mentioned. I'm looking forward to seeing how much more efficient our pipelines will be.

> Another solution that we're considering involves adding a "next datum" API call which could be used inside of user code.

If I'm allowed to weigh in on a solution at all, I would prefer not to have to incorporate an API call in my pipeline code. The architecture and design of Pachyderm so far has been so well thought out that it allows my pipeline code to be agnostic to the fact that it's being executed inside Pachyderm. The only thing the pipeline code needs to assume is that input data will show up in certain input directories and output data needs to written to a specific output directory. In other words, the file system is the interface between Pachyderm and pipeline code, and the pipeline code just needs to follow certain conventions around how the file system is used. To me, this feels like a very clean interface, but adding the "next datum" API pokes a hole in that interface.

One thing that's great about the file system interface for my specific use case is that it allows me to develop and unit test my pipelines locally and outside of a Pachyderm platform. In development or automated unit tests, I can manipulate the file system just as well as Pachyderm by creating temporary directory structures that mimic the Pachyderm convention. I can then run my pipelines directly on those directories and validate the output, all without having to deploy a pipeline spec. Now, the fidelity between how my pipelines run in testing/development vs. production is much higher. If the "next datum" API was implemented, it would be more challenging for me to mock that out in testing. The proposed/alternative solution (i.e. restructuring the `/pfs` dir) is consistent with the current file system interface, and therefore my development and testing workflows could easily adapt.

**Comment by jdoliner** [comment]
@claytonlemons of course you're allowed to weigh in on the solution, we always welcome community feedback. This particular issue is one we've iterated on a bunch so let me explain how we arrived at this solution. Initially we liked a solution like the one you describe where we restructure the `/pfs` directory. However this had a few issues that made us hesitant:

- it forces users to restructure their code to read from multiple directories and to get the best performance make their code run concurrently
- it doesn't totally fix the performance issues we're targeting, it more just changes the constants. You still have to exec a user process for every `n` datums. `n` is configurable and can be large, but there are tradeoffs, too small and you're not really solving the originally problem. Too large and the size of the data becomes a worry, all of the datums must be downloaded before the user code can be started, whereas in the current model we interleave downloading, processing and uploading. Furthermore, all of the datums, and output from processing, must be stored on a node at the same time which increases the amount of diskspace the system as a whole requires.
- it requires commits to be large to work because we can't batch datums across commits.

All of these issues come from the fundamental issue that there's an implicit api for declaring your datums processed which is the process exiting, so you have to exit at some point. The only way to fix that problem is to introduce an alternative api which does the same thing which is what `next datum` does. Although this does introduce a pachyderm specific api into the users container we think that's worth the complexity because that's the only thing you need to do when you want to avoid the boot time of your code, call `next datum` loop back to the beginning of your code. We find that to be much simpler than needing to handle a different `/pfs` structure. Furthermore we recently changed our spout interface to inject `pachctl` into the user container and have found that to be an effective and fairly easy to use api for people and we feel this interface is similar.
